# Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
# Use subject to license terms.
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, and/or sell copies of the Software, and to permit persons
# to whom the Software is furnished to do so, provided that the above
# copyright notice(s) and this permission notice appear in all copies of
# the Software and that both the above copyright notice(s) and this
# permission notice appear in supporting documentation.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT
# OF THIRD PARTY RIGHTS. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
# HOLDERS INCLUDED IN THIS NOTICE BE LIABLE FOR ANY CLAIM, OR ANY SPECIAL
# INDIRECT OR CONSEQUENTIAL DAMAGES, OR ANY DAMAGES WHATSOEVER RESULTING
# FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
# WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
# 
# Except as contained in this notice, the name of a copyright holder
# shall not be used in advertising or otherwise to promote the sale, use
# or other dealings in this Software without prior written authorization
# of the copyright holder.

6484245: auto-config: radeon: misc fixes to radeon driver - revision to 6436994

diff -ur src.orig/radeon_mergedfb.c src/radeon_mergedfb.c
--- src.orig/radeon_mergedfb.c	Mon Jan  8 14:26:27 2007
+++ src/radeon_mergedfb.c	Mon Jan  8 14:28:26 2007
@@ -212,19 +212,15 @@
     mode->VSyncEnd += dy;
     mode->VTotal += dy;
 
-    /* This is needed for not generating negative refesh rates in xrandr with the
-       faked DotClock below
-     */
-    if (!(mode->VRefresh))
-        mode->VRefresh = mode->Clock * 1000.0 / mode->HTotal / mode->VTotal;
-
-     /* Provide a sophisticated fake DotClock in order to trick the vidmode
+     /* Provide a fake VRefresh/DotClock in order to trick the vidmode
       * extension to allow selecting among a number of modes whose merged result
       * looks identical but consists of different modes for CRT1 and CRT2
       */
-    mode->Clock = (((i->Clock >> 3) + i->HTotal) << 16) | ((j->Clock >> 2) + j->HTotal);
-    mode->Clock ^= ((i->VTotal << 19) | (j->VTotal << 3));
+    mode->VRefresh = (float)((i->Clock * 1000.0 / i->HTotal / i->VTotal) * 100 +
+	(j->Clock * 1000.0 / j->HTotal / j->VTotal));
 
+    mode->Clock = (int)(mode->VRefresh * 0.001 * mode->HTotal * mode->VTotal);
+
     if( ((mode->HDisplay * ((pScrn->bitsPerPixel + 7) / 8) * mode->VDisplay) > 
 	(pScrn->videoRam * 1024)) ||
         (mode->HDisplay > 8191) ||
